### PR TITLE
 Document triage of failed merge commits in operator manual

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
No issue.

Author

- [ ] ~~PR title references issue~~
- [ ] ~~PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>~~
- [ ] ~~Title of main commit references issue~~
- [ ] ~~PR is connected to Zenhub issue and description links to issue~~

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved ~linked issue~ PR to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Review by amarjandu — APPROVED** [review]
approved based on conversation https://ucsc-gi.slack.com/archives/C705Y6G9Z/p1632840435386600

**Comment by noah-aviel-dove** [comment]
@amarjandu I pushed a second commit based on parking lot to avoid opening a second PR.

**Review by amarjandu — APPROVED** [review]
this was agreed on during parking lot.

**Comment by hannes-ucsc** [comment]
We've decided in last retro that I review even these PR.

File: OPERATOR.rst [hunk 1]
```diff
@@ -93,6 +94,19 @@ Triaging ``sandbox`` failures
   discretion on whether or not to open a new ticket, it is important to record
   unusual failures to determine their eventual frequency.
 
+Triaging merge failures on ``dev`` and ``prod``
+"""""""""""""""""""""""""""""""""""""""""""""""
+
+If the build for a merge commit fails on a stable deployment, the operator must
```
[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply OPERATOR.rst:100]: I think this belongs in the `Reindexing` section instead, as there may be failed messages in the queues regardless of whether the build passed or not.

File: OPERATOR.rst [hunk 2]
```diff
@@ -93,14 +94,36 @@ Triaging ``sandbox`` failures
...
 ^^^^^^^^^^
 
 The operator must check the status of the queues after every reindex for
 failures. Use ``python scripts/manage_queues.py`` to identify any failed
-messages. If failed messages are found
+messages. If failed messages are found, use ``python scripts/manage_queues.py``
+to
+
+- dump the failed messages to JSON file(s)
+- clear the fail queues so they are empty for the next reindex
```
Comment (hannes-ucsc) [inline OPERATOR.rst:121]: This can be one step, using `--delete` with dump.

File: OPERATOR.rst [hunk 3]
```diff
@@ -93,14 +94,36 @@ Triaging ``sandbox`` failures
...
+- feed the failed messages back into their respective ``_retry`` queues
```
Comment (hannes-ucsc) [inline OPERATOR.rst:122]:
```suggestion
- feed the failed messages back into their respective ``_retry`` queues, not the non-retry queues (!)
```

or some other, less awkward way to emphasize this.

File: OPERATOR.rst [hunk 4]
```diff
@@ -93,14 +94,36 @@ Triaging ``sandbox`` failures
...
+
 .. _committing the changes separately: https://github.com/DataBiosphere/azul/issues/2899#issuecomment-804508017
 
 Reindexing
 ^^^^^^^^^^
 
 The operator must check the status of the queues after every reindex for
 failures. Use ``python scripts/manage_queues.py`` to identify any failed
-messages. If failed messages are found
+messages. If failed messages are found, use ``python scripts/manage_queues.py``
```
Comment (hannes-ucsc) [inline OPERATOR.rst:117]: Should discriminate between notifcations and tallies. I think the best sequence would be notifications until the set of failed messages stabilizes, then repeat with tallies.

File: OPERATOR.rst [hunk 5]
```diff
@@ -93,14 +94,36 @@ Triaging ``sandbox`` failures
...
+to
+
+- dump the failed messages to JSON file(s)
+- clear the fail queues so they are empty for the next reindex
+- feed the failed messages back into their respective ``_retry`` queues
+
+This may cause the previously failed messages to succeed. Repeat this procedure
+until no progress towards success is being made, i.e., the fail queues are empty
+or no previosuly failed messages succeeded. Finally
```
Comment (hannes-ucsc) [inline OPERATOR.rst:126]: Should state the invariant post condition that all queues must be empty and that any permanent failures are tracked in tickets.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by noah-aviel-dove** [comment]
@hannes-ucsc  Build is red due to https://github.com/DataBiosphere/azul/issues/3603

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Build must be green.

File: OPERATOR.rst [hunk 6]
```diff
@@ -93,14 +94,46 @@ Triaging ``sandbox`` failures
   discretion on whether or not to open a new ticket, it is important to record
   unusual failures to determine their eventual frequency.
 
+Triaging GitLab build failures on ``dev`` and ``prod``
+""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+If a GitLab build fails on a stable deployment, the operator must evaluate the
```
Comment (hannes-ucsc) [inline OPERATOR.rst:100]:
```suggestion
If a GitLab build fails on a main deployment, the operator must evaluate the
```

File: OPERATOR.rst [hunk 7]
```diff
@@ -93,14 +94,46 @@ Triaging ``sandbox`` failures
...
 
 The operator must check the status of the queues after every reindex for
 failures. Use ``python scripts/manage_queues.py`` to identify any failed
-messages. If failed messages are found
+messages. If failed messages are found, use ``python scripts/manage_queues.py``
+to
+
+- dump the failed notifications to JSON file(s), using ``--delete`` to
+  simultaneously clear the ``notifications_fail`` queue
+- force-feed the failed notifications back into the ``notifications_retry``
```
Comment (hannes-ucsc) [inline OPERATOR.rst:122]:
```suggestion

- force-feed the failed notifications back into the ``notifications_retry``
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**